lasagne/layers/shape.py

```diff
@@ -375,9 +375,11 @@ def get_output_shape_for(self, input_shape):
         output_shape = list(input_shape)
         if isinstance(self.slice, int):
             del output_shape[self.axis]
-        else:
+        elif input_shape[self.axis] is not None:
             output_shape[self.axis] = len(
                 range(*self.slice.indices(input_shape[self.axis])))
+        else:
+            output_shape[self.axis] = None
```

Comment:
You could omit the last two lines, `output_shape[self.axis]` is already None if `input_shape[self.axis]` is None. (But you can answer "explicit is implicit" and I'm with it :). As you prefer.)